### PR TITLE
Skill Trees Demo

### PR DESCRIPTION
**Pull request by tgao564**
We have added the skilltrees_demo.c file in the src/skill/examples folder as our demo file for the final presentation. The file includes an implementation for a demo that showcases the new effects and leveling implementation done by the skill trees team this quarter. This PR relies on code in PR #1110 and therefore that PR needs to be approved before this one can.

### CODE REVIEW

**Comment by rohangupta2806** [comment]
PR 1110 has been merged.  This PR can now be reviewed.

**Review by mariakravtsova — APPROVED** [review]
Good demo!

File: src/skilltrees/examples/skilltrees_demo.c [hunk 1]
```diff
@@ -0,0 +1,378 @@
...
+}
+
+void create_player_skill(chiventure_ctx_t* ctx)
+{
+    /*Creating an effect with a hardcoded value of 100 */
+    char* stats_to_change[] = {"max_health", "current_health"};
+    double mods[] = {100, 100};
+    int durations[] = {5, 5};
+    player_stat_effect_t* health_boost = define_player_stat_effect("health boost", stats_to_change, mods, durations, 2, ctx);
+    if(health_boost == NULL)
```
Comment (firatciftci) [inline src/skilltrees/examples/skilltrees_demo.c:141]: Make sure to add a space character between `if` and `(`. Also, make sure to always add curly brackets around if statements, with proper indentation for the code block inside the if statement.

File: src/skilltrees/examples/skilltrees_demo.c [hunk 2]
```diff
@@ -0,0 +1,378 @@
...
+    int i = 1;
+    while ((skill->sid != sid)&&(i<=1))
+    {
+        skill_node = player->player_class->skilltree->nodes[i];
+        i +=1;
+        skill = skill_node->skill;
+    }
+    
+    /* Check the level */
+    if(player->level<skill_node->prereq_level)
```
Comment (firatciftci) [inline src/skilltrees/examples/skilltrees_demo.c:180]: Make sure to add a space character between `if` and `(`.

File: src/skilltrees/examples/skilltrees_demo.c [hunk 3]
```diff
@@ -0,0 +1,378 @@
...
+    skill_t* skill = skill_node->skill;
+    
+    /*Find the correct skill */
+    int i = 1;
+    while ((skill->sid != sid)&&(i<=1))
+    {
+        skill_node = player->player_class->skilltree->nodes[i];
+        skill = skill_node->skill;
+    }
+    if(skill->sid != sid)
```
Comment (firatciftci) [inline src/skilltrees/examples/skilltrees_demo.c:205]: Make sure to add a space character between `if` and `(`.

File: src/skilltrees/examples/skilltrees_demo.c [hunk 4]
```diff
@@ -0,0 +1,378 @@
...
+    }
+    /* Execute the effect */
+    skill_execute(skill_node->skill, ctx);
+    return SUCCESS;
+}
+
+char* add_player_stat_operation(char* tokens[TOKEN_LIST_SIZE], chiventure_ctx_t* ctx)
+{
+    int check = add_skill_to_player(ctx, 0);
+    if(check == FAILURE)
```
Comment (firatciftci) [inline src/skilltrees/examples/skilltrees_demo.c:217]: Make sure to add a space character between `if` and `(`.

File: src/skilltrees/examples/skilltrees_demo.c [hunk 5]
```diff
@@ -0,0 +1,378 @@
...
+    }
+}
+
+void create_attr_skill(chiventure_ctx_t* ctx)
+{
+    attribute_value_t mod;
+    mod.bool_val = false;
+    enum attribute_tag att_tag = BOOLE;
+    item_t* dragon = get_item_in_hash(ctx->game->all_items, "DRAGON");
+    if(dragon == NULL)
```
Comment (firatciftci) [inline src/skilltrees/examples/skilltrees_demo.c:234]: Make sure to add a space character between `if` and `(`.

File: src/skilltrees/examples/skilltrees_demo.c [hunk 6]
```diff
@@ -0,0 +1,378 @@
...
+char* create_attr_skill_operation(char* tokens[TOKEN_LIST_SIZE], chiventure_ctx_t* ctx)
+{
+    create_attr_skill(ctx);
+    return "Created Attribute modifying skill!";
+}
+
+char* add_attr_skill_operation(char* tokens[TOKEN_LIST_SIZE], chiventure_ctx_t* ctx)
+{
+    int check = add_skill_to_player(ctx, 1);
+    if(check == FAILURE)
```
Comment (firatciftci) [inline src/skilltrees/examples/skilltrees_demo.c:256]: Make sure to add a space character between `if` and `(`.

File: src/skilltrees/examples/skilltrees_demo.c [hunk 7]
```diff
@@ -0,0 +1,378 @@
...
+    {
+        return "Added skill!";
+    }
+}
+
+char* execute_attr_operation(char* tokens[TOKEN_LIST_SIZE], chiventure_ctx_t* ctx)
+{
+    int check = execute_skill(ctx, 1);
+    item_t* dragon = get_item_in_hash(ctx->game->all_items, "DRAGON");
+    if(dragon == NULL)
```
Comment (firatciftci) [inline src/skilltrees/examples/skilltrees_demo.c:270]: Make sure to add a space character between `if` and `(`. Also, make sure to always add curly brackets around if statements, with proper indentation for the code block inside the if statement.

File: src/skilltrees/examples/skilltrees_demo.c [hunk 8]
```diff
@@ -0,0 +1,378 @@
...
+    fprintf(stderr, "NO DRAGON");
+    bool is_alive = dragon->attributes->attribute_value.bool_val;
+    if(is_alive == true)
```
Comment (firatciftci) [inline src/skilltrees/examples/skilltrees_demo.c:273]: Make sure to add a space character between `if` and `(`. Also, make sure to always add curly brackets around if statements, with proper indentation for the code block inside the if statement.

File: src/skilltrees/examples/skilltrees_demo.c [hunk 9]
```diff
@@ -0,0 +1,378 @@
...
+    print_to_cli(ctx, "Passive Skills:");
+    if (!inventory->num_passive) {
+        print_to_cli(ctx, "You have no passive skills.");
+    } else {
+        for (i = 0; i < inventory->num_passive; i++) {
+            sprintf(description, "%s: Level %u", inventory->passive[i]->name,
+                    inventory->passive[i]->level);
+            print_to_cli(ctx, description);
+        }
+    }
```
Comment (firatciftci) [inline src/skilltrees/examples/skilltrees_demo.c:343]: This file seems to be using a new-line for opening curly brackets, so I think you should move the `{` at the end of the if, else, and for statements in this section to a new-line just below each one.

File: src/skilltrees/examples/skilltrees_demo.c [hunk 10]
```diff
@@ -0,0 +1,378 @@
...
+ *
+ * Returns:
+ *  - An empty string (current_skills_as_strings() prints all that is necessary)
+ */
+char* skills_operation(char *tokens[TOKEN_LIST_SIZE], chiventure_ctx_t* ctx){
+    current_skills_as_strings(ctx);
+    return "";
+}
+
+void main(){
```
Comment (firatciftci) [inline src/skilltrees/examples/skilltrees_demo.c:362]: Make sure to add a space character between `)` and `{`.

**Review by firatciftci — CHANGES_REQUESTED** [review]
I have requested some stylistic changes.

**Comment by rohangupta2806** [comment]
I made all the required changes.  Should be good to merge now.

**Review by firatciftci — APPROVED** [review]
Everything looks good to me!

**Comment by rohangupta2806** [comment]
Closing comment:  The demo for skilltrees has been merged.  The demo can be built by going to the build/src/skilltrees/examples folder and doing make, and then running skilltrees_demo.  We have added a number of custom commands that let us showcase the functionality of skilltrees and the new effects, along with prerequisite levels.

**Comment by firatciftci** [comment]
Issue Score: ✔️++

Comments:
Good work!

**Pull request merged**